src/assets/RustAsset.js

```diff
@@ -0,0 +1,93 @@
+const url = require('url');
+const path = require('path');
+const toml = require('toml');
+const {exec} = require('child-process-promise');
+
+const fs = require('../utils/fs');
+const JSAsset = require('./JSAsset');
+const localRequire = require('../utils/localRequire');
+
+const rustTarget = `wasm32-unknown-emscripten`;
+
+class RustAsset extends JSAsset {
+  collectDependencies() {
+    // Do nothing. Dependencies are collected by cargo :).
+  }
+
+  async parse(code) {
+    this.invalidateBundle();
+    this.invalidate();
+    this.dependencies.clear();
+    const {name, encoding, options} = this;
+    const release = process.env.NODE_ENV === 'production';
+    const cmd = `wargo build ${release ? ' --release' : ''}`;
+
+    const projectDir = path.join(path.dirname(name), '..');
+    const cargoFile = await fs.readFile(
+      path.join(projectDir, 'Cargo.toml'),
+      encoding
+    );
+    const packageName = toml.parse(cargoFile.toString()).package.name;
+
+    const outDir = path.join(
+      projectDir,
+      'target',
+      rustTarget,
+      release ? 'release' : 'debug'
+    );
+    const depsDir = path.join(outDir, 'deps');
+
+    const outFile = path.join(outDir, `${packageName}.js`);
+
+    await exec(cmd, {cwd: projectDir});
+
+    const out = await fs.readFile(outFile, encoding);
+    const deps = await fs.readdir(path.join(outDir, 'deps'));
+
+    const wasmFile = deps.find(f => /\.wasm$/.test(f));
+
+    if (!wasmFile) {
+      throw new Error('No wasm file found', null);
+    }
+
+    const generatedName = this.addURLDependency(path.join(depsDir, wasmFile));
+
+    const wasmBinaryFile = url.resolve(
+      path.join(options.publicURL, generatedName),
+      ''
+    );
+
+    const Module = {
+      // Path in the built project to the wasm file
+      wasmBinaryFile,
+      // Indicates that we are NOT running in node, despite 'require' being defined
+      ENVIRONMENT: 'WEB'
+    };
+
+    if (options.fromHtml) {
```

Comment:
Why is it different when loaded from HTML?

Reply:
the difference is that with this approach you can use rust as a standalone executable, without javascript.

I got the idea from this [article](https://hoverbear.org/2017/04/06/the-path-to-rust-on-the-web/)